### PR TITLE
Safebooru Waifu Gadget

### PR DESCRIPTION
**Pull request by lucastarche**
This gadget is of vital importance to the project, given the moral support it will give to both its users and developers.

It queries a random image from safebooru, and displays it to the user.
Don't believe the effectiveness? Just look at this one simple comparison:

unknown

While the first image looks bland and boring, the sole presence of an anime girl brightens the environment, and makes it friendly and approachable.

### CODE REVIEW

File: gadgets/safebooru-waifu/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        "Safebooru Waifu"
+    }
+
+    fn make_gadget(
+        &self,
+        network_runtime: &utils::NetworkRuntime,
+        egui_ctx: &egui::Context,
+    ) -> Box<dyn Gadget> {
+        let waifu_gadget = WaifuGadget {
+            image: Arc::new(Mutex::new(None)),
```
Comment (RealKC) [inline gadgets/safebooru-waifu/src/lib.rs]: What do you think of doing this instead:
```suggestion
            image: Arc::default()
```

File: gadgets/safebooru-waifu/src/safebooru.rs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+pub async fn query_random_image() -> anyhow::Result<RetainedImage> {
+    let response = reqwest::get(QUERY_URL).await?.text().await?;
+    let response: SafebooruResponse = serde_json::from_str(&response)?;
+    let image_response = reqwest::get(response.file_url).await?;
+
+    let image = RetainedImage::from_image_bytes(
+        image_response.url().to_string(),
+        &image_response.bytes().await?,
+    )
+    .map_err(|e| Error::new(std::io::ErrorKind::Other, e))?;
```
Comment (RealKC) [inline gadgets/safebooru-waifu/src/safebooru.rs]: Why's this needed?

**Review by RealKC — CHANGES_REQUESTED** [review]

File: gadgets/safebooru-waifu/src/safebooru.rs (hunk 2 above)
Comment (lucastarche) [reply gadgets/safebooru-waifu/src/safebooru.rs]: `from_image_bytes()` returns a `Result<RetainedImage, String>`, and `String` does not implement `StdError` sadly

File: gadgets/safebooru-waifu/src/lib.rs (hunk 1 above)
Comment (lucastarche) [reply gadgets/safebooru-waifu/src/lib.rs]: Sure, I don't mind it

File: gadgets/safebooru-waifu/src/safebooru.rs (hunk 2 above)
Comment (RealKC) [reply gadgets/safebooru-waifu/src/safebooru.rs]: That's an unfortunate API, but I see now. I'm not happy about us using io::Error for this though, what about something like:

```rs
// RetainedImage::from_image_bytes returns a Result<RetainedImage, String>, so we make our own error for anyhow interop
struct RetainedImageError(String);

impl Display for RetainedImageError { ... }
impl Error for RetainedImageError {}
```

Comment (lucastarche) [reply gadgets/safebooru-waifu/src/safebooru.rs]: We should probably have that somewhere in utils/ but I agree mapping a String to an io::Error is a hacky way to go about it

**Review by RealKC — APPROVED** [review]

**Pull request merged**